### PR TITLE
[8.x] Adds new `RequestSent` and `ResponseReceived` events to the HTTP Client

### PR DESCRIPTION
**Pull request by lukeraymonddowning**
Hey guys,

This PR adds two new events, `RequestSent` and `ResponseReceived`, which are dispatched by the Http Client.

## `RequestSent`

The `RequestSent` event is fired when any request is made when calling any of the following Http Client methods: 'get', 'head', 'post', 'put' and 'delete'. It receives the HTTP method, URL, and any sent options.

## `ResponseReceived`

The `ResponseReceived` event is fired when a successful response is returned from the Http Client. It receives the `Response` instance.

## What are the benefits?

These two events allow a 3rd party package to listen for and react to HTTP Client requests. I can see it being used for debugging, logging, performance monitoring, service downtime analysis and more.

I decided to contribute this when I went to create a PR for the `spatie/laravel-ray` package to allow monitoring of Http Client requests only to realise there wasn't a way to do so.

As always, thank you very much for all the hard work and for considering my PR.

Regards,
Luke

### CODE REVIEW

**Comment by taylorotwell** [comment]
Renamed to `RequestSending` (it is dispatched slightly before the request is actually sent), and moved its location to `sendRequest` method in case of promise usage. Otherwise, the event would have fired before the request was actually sending, which could have been confusing.

Thoughts?

File: src/Illuminate/Http/Client/Factory.php [hunk 1]
```diff
@@ -85,10 +93,13 @@ class Factory
     /**
      * Create a new factory instance.
      *
+     * @param  \Illuminate\Contracts\Events\Dispatcher  $dispatcher
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/Factory.php:96]: missing |null

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by GrahamCampbell** [comment]
Why not include the request object in the first event?

**Comment by taylorotwell** [comment]
I agree having the full Illuminate\Http\Client\Request object would be useful in both events to be honest.

**Comment by lukeraymonddowning** [comment]
@taylorotwell agreed. Also agreed on the received event having the response object. I'll refactor when I'm back at my keyboard.

**Comment by lukeraymonddowning** [comment]
Okay @taylorotwell @GrahamCampbell, I've moved the firing of the `RequestSending` event to a `beforeSendingCallback` so that we can include the `Request` object. 

I've also made the `ResponseReceived` event accept the `Request` too, so that you have both the request and response in that event. I've done this by setting the request in a property on the `PendingRequest`, which seemed the cleanest way to achieve this. Can you see any issues with that approach, or does that seem good to you?

**Comment by lukeraymonddowning** [comment]
Also @taylorotwell, completely understand if not (because seeing 'Request' as the first word makes sense in one way), but would it be worth considering `SendingRequest` as the event name, just so that it reads better as a phrase? Not precious, but thought it was worth raising as a question.

File: src/Illuminate/Http/Client/Factory.php (hunk 1 above)
Comment (lukeraymonddowning) [reply src/Illuminate/Http/Client/Factory.php]: Added 👍

File: tests/Http/HttpClientTest.php [hunk 2]
```diff
@@ -21,7 +26,7 @@
 class HttpClientTest extends TestCase
 {
     /**
-     * @var \Illuminate\Http\Client\Factory
+     * @var Factory
```
Comment (driesvints) [inline tests/Http/HttpClientTest.php:29]: Revert this change

File: tests/Http/HttpClientTest.php [hunk 3]
```diff
@@ -367,7 +372,7 @@ public function testWithCookies()
 
         $this->assertCount(1, $response->cookies()->toArray());
 
-        /** @var \GuzzleHttp\Cookie\CookieJarInterface $responseCookies */
+        /** @var CookieJarInterface $responseCookies */
```
Comment (driesvints) [inline tests/Http/HttpClientTest.php:375]: Revert this change

File: tests/Http/HttpClientTest.php (hunk 2 above)
Comment (lukeraymonddowning) [reply tests/Http/HttpClientTest.php]: Sorry Dries. Not a clue why they changed, must have been my IDE

Comment (driesvints) [reply tests/Http/HttpClientTest.php]: @lukeraymonddowning no worries :)
